core/src/smartcontracts/isi/triggers.rs

```diff
@@ -34,32 +14,61 @@ pub mod isi {
 
     use super::{super::prelude::*, *};
 
-    impl<W: WorldTrait> Execute<W> for Register<Trigger> {
+    impl<W: WorldTrait> Execute<W> for Register<Trigger<FilterBox>> {
         type Error = Error;
 
         #[metrics(+"register_trigger")]
+        #[allow(clippy::expect_used)]
         fn execute(
             self,
             _authority: <Account as Identifiable>::Id,
             wsv: &WorldStateView<W>,
         ) -> Result<(), Self::Error> {
             let new_trigger = self.object;
 
-            if new_trigger.action.occurs_exactly_at_time()
-                && !matches!(&new_trigger.action.repeats, Repeats::Exactly(1))
-            {
-                return Err(Error::Math(MathError::Overflow));
+            if !new_trigger.action.mintable() {
+                match &new_trigger.action.repeats {
+                    Repeats::Exactly(action) if action.get() == 1 => (),
+                    _ => {
+                        return Err(Error::Math(MathError::Overflow));
+                    }
+                }
             }
 
             wsv.modify_triggers(|triggers| {
                 let trigger_id = new_trigger.id.clone();
-                triggers.add(new_trigger)?;
+                match &new_trigger.action.filter {
+                    FilterBox::Data(_) => {
+                        triggers.add_data_trigger(new_trigger.try_into().expect(
```

Comment:
At this point a panic is no different from an error message. What you could do is return the error, and use `expect` at the calling sites, but IMO, Iroha itself should not contain `expect` unless something is provably `Infallible`